libr/anal/p/anal_arm_cs.c

```diff
@@ -3108,32 +3125,67 @@ static void set_opdir(RAnalOp *op) {
         }
 }
 
-static void op_fillval(RAnalOp *op , csh handle, cs_insn *insn, int bits) {
-	static RRegItem reg;
-	switch (op->type & R_ANAL_OP_TYPE_MASK) {
-	case R_ANAL_OP_TYPE_LOAD:
-		op->src[0] = r_anal_value_new ();
-		ZERO_FILL (reg);
-		op->src[0]->reg = &reg;
+static void set_src_dst(RAnalValue *val, csh *handle, cs_insn *insn, int x, int bits) {
+	cs_arm_op armop = INSOP (x);
+	cs_arm64_op arm64op = INSOP64 (x);
+	if (bits == 64) {
+		parse_reg64_name (&base_regs[x], &regdelta_regs[x], *handle, insn, x);
+	} else {
+		parse_reg_name (&base_regs[x], &regdelta_regs[x], *handle, insn, x);
+	}
+	switch (armop.type) {
+	case ARM_OP_REG:
+		break;
+	case ARM_OP_MEM:
 		if (bits == 64) {
-			parse_reg64_name (op->src[0]->reg, handle, insn, 1);
-			op->src[0]->delta = MEMDISP64(1);
+			val->delta = arm64op.mem.disp;
 		} else {
-			parse_reg_name (op->src[0]->reg, handle, insn, 1);
-			op->src[0]->delta = MEMDISP(1);
+			val->mul = armop.mem.scale;
+			val->delta = armop.mem.disp;
 		}
+		val->regdelta = &regdelta_regs[x];
+		break;
+	default:
+		break;
+	}
+	val->reg = &base_regs[x];
+}
+
+static void create_src_dst(RAnalOp *op) {
+	op->src[0] = r_anal_value_new ();
+	op->src[1] = r_anal_value_new ();
+	op->src[2] = r_anal_value_new ();
+	op->dst = r_anal_value_new ();
+	ZERO_FILL (base_regs[0]);
+	ZERO_FILL (base_regs[1]);
+	ZERO_FILL (base_regs[2]);
+	ZERO_FILL (base_regs[3]);
+	ZERO_FILL (regdelta_regs[0]);
+	ZERO_FILL (regdelta_regs[1]);
+	ZERO_FILL (regdelta_regs[2]);
+	ZERO_FILL (regdelta_regs[3]);
+}
+
+static void op_fillval(RAnalOp *op , csh handle, cs_insn *insn, int bits) {
+	create_src_dst (op);
+	switch (op->type & R_ANAL_OP_TYPE_MASK) {
+	case R_ANAL_OP_TYPE_LOAD:
+	case R_ANAL_OP_TYPE_MOV:
+	case R_ANAL_OP_TYPE_SUB:
+	case R_ANAL_OP_TYPE_ADD:
+	case R_ANAL_OP_TYPE_AND:
+	case R_ANAL_OP_TYPE_XOR:
+	case R_ANAL_OP_TYPE_MUL:
+	case R_ANAL_OP_TYPE_CMP:
+	case R_ANAL_OP_TYPE_OR:
+		set_src_dst (op->src[2], &handle, insn, 3, bits);
+		set_src_dst (op->src[1], &handle, insn, 2, bits);
+		set_src_dst (op->src[0], &handle, insn, 1, bits);
+		set_src_dst (op->dst, &handle, insn, 0, bits);
 		break;
 	case R_ANAL_OP_TYPE_STORE:
-		op->dst = r_anal_value_new ();
-		ZERO_FILL (reg);
-		op->dst->reg = &reg;
-		if (bits == 64) {
-			parse_reg64_name (op->dst->reg, handle, insn, 1);
-			op->dst->delta = MEMDISP64(1);
-		} else {
-			parse_reg_name (op->dst->reg, handle, insn, 1);
-			op->dst->delta = MEMDISP(1);
-		}
+		set_src_dst (op->dst, &handle, insn, 1, bits);
```

Comment:
uhm isnt this 0, and 1, instead of 1 and 0?

Reply:
I suppose no; str in ARM has source as first operand as far as I know.